feat: Dynamic light sleep support for ESP32 platform

m1nl opened the pull request:
fixes #6660 

I've finally implemented the Feature Request I created a few months ago. Since then, I've been testing the dynamic light-sleep functionality and haven't encountered any issues.

In the initial version, legacy light-sleep support with the default Arduino framework could lead to problems after my changes. To address this, I had to rewrite and refactor parts of the code to ensure compatibility with builds compiled using the default PlatformIO Arduino framework.

I tested the updated implementation on a Heltec V3 board in the following scenarios:
- default Arduino, bluetooth, light-sleep disabled
- default Arduino, bluetooth, light-sleep enabled
- default Arduino, bluetooth, light-sleep enabled, force disable support for EXT wake-up
- customized Arduino, bluetooth, light-sleep disabled
- customized Arduino, bluetooth, light-sleep enabled
- customized Arduino, bluetooth, light-sleep enabled, force disable support for EXT wake-up
- customized Arduino, bluetooth disabled, light-sleep enabled
- customized Arduino, WiFi enabled, MQTT enabled, light-sleep enabled

There are a few topics that need discussion:
1) Power Management (PM) Support for ESP32 requires a custom variant of the Arduino framework. I’ve forked it from Espressif’s GitHub repository and am currently hosting it under my personal GitHub account. If these changes are accepted, I’d like to move the repository under the `meshtastic` organization.
2) In my opinion, re-enabling Bluetooth on the ESP32 without restarting is not possible, so NB PowerFSM state seems redundant. I'm open to restoring it if necessary, but I’d appreciate guidance on the intended logic behind this state.
3) I updated the `platformio.ini` files for the hardware variants I was able to test. I shared customized builds with others, who confirmed they were working as expected.
4) I updated the `variant.h` files for the hardware variants which do have 32kHz oscillator, according to Espressif docs, it helps with Bluetooth stability when dynamic light-sleep is enabled

Links to repositiories mentioned in 1) :
- https://github.com/m1nl/arduino-esp32
- https://github.com/m1nl/esp32-arduino-lib-builder

## 🤝 Attestations

- [x] I have tested that my proposed changes behave as described.
- [x] I have tested that my proposed changes do not cause any obvious regressions on the following devices:
  - [x] Heltec (Lora32) V3
  - [ ] LilyGo T-Deck
  - [x] LilyGo T-Beam
  - [ ] RAK WisBlock 4631
  - [x] Seeed Studio T-1000E tracker card
  - [ ] Other (please specify below)

m1nl committed:
refactor sleep and PowerFSM to allow dynamic light sleep

m1nl committed:
prevent light sleep when modules are busy for more than one timer tick

m1nl committed:
dynamically adjust measurement interval in power management

m1nl committed:
switch heltec_v3 variant to esp-idf sdk with power management enabled

m1nl committed:
move sleep wake timeout to PhoneAPI

m1nl committed:
use events to stay in idle state

m1nl committed:
add assertion for gpio_wakeup_enable calls

m1nl committed:
allow light sleep when MQTT is enabled

m1nl committed:
improve logging

fix logging

make log entry more consistent

m1nl committed:
prevent zero-byte read from serial causing too frequent wakeups

m1nl committed:
do not detach button interrupt when going into light sleep

m1nl committed:
remove stateNB, improve regular (non-dynamic) light-sleep handling

m1nl committed:
move feature flags

m1nl committed:
implement lower cap on power thread interval

m1nl committed:
make some variables static, refactor GPIO reset logic

m1nl committed:
another refactor to better handle variants without dynamic light sleep

m1nl committed:
fix gpio reset (rtc gpios not supported with esp32c6)

m1nl committed:
exit light-sleep state when bluetooth pairing is received

fifieldt commented:
This is a ton of work, thank you very much

fifieldt commented:
If you have time .... I "maintain" the GPS code - just wondering what the impact might be there or whether we might need further extensions?

m1nl commented:
AFAIK dynamic light-sleep in ESP32 platform makes the chip light-sleep when FreeRTOS waits for an event or there is an explicit vTaskDelay call (=~ sleep in Arduino). So most of the functionality shouldn't be impacted as Arduino runs all the code in one loop, which waits different intervals depending on the type of thread (using vTaskDelay under the hood). However, for critical modules or core functionality I added and explicit call to PowerFSM.trigger to exit LS state and allow for at least 300ms of uninterrupted processing time; another approach is to use "preflightSleepObserver" which prevents entering LS state if there is some work pending (like TX queue is not empty). I took a look at GPS code and I think we should exit LS state when we change state to GPS_ACTIVE and prevent entering LS as long as the state doesn't change. I'll take a look at the code and propose some changes soon.

m1nl commented:
If there is a chance to have this feature merged, I'd appreciate getting some feedback on the additional repos to be hosted under `meshtastic` Github organization. As mentioned above, default Arduino framework provided by Espressif doesn't enable PM functionality at all (which is really bad IMO).

m1nl committed:
do not enter light-sleep when GPS is in GPS_ACTIVE state

m1nl commented:
@fifieldt changes committed; device should not go into light-sleep when GPS position is requested (so module is in GPS_ACTIVE state). I'm unable to test as I'm on vacation right now.

m1nl commented:
hi,

1) Bumping my last question

I'd appreciate getting some feedback on hosting additional repos under meshtastic Github organization.  Default Arduino framework provided by Espressif doesn't enable PM functionality at all. I'm happy to move content of my repos to new ones. There are two repos to be created - arduino-esp32 and esp32-arduino-lib-builder

2) I took a look at the code and it reminded me I did change the WiFi code a bit - I'm enabling power saving by default (regardless of powersaving setting in module options); I think I should revoke that change to ensure compatibility with older versions; let me know what you think

m1nl committed:
keep CPU at full speed when not power saving and WiFi is enabled

m1nl commented:
> ```
> /work/.platformio/packages/toolchain-xtensa-esp32/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pio/build/heltec-v2_1/firmware.elf section `.iram0.text' will not fit in region `iram0_0_seg'
> /work/.platformio/packages/toolchain-xtensa-esp32/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: IRAM0 segment data does not fit.
> /work/.platformio/packages/toolchain-xtensa-esp32/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: region `iram0_0_seg' overflowed by 776 bytes
> collect2: error: ld returned 1 exit status
> ```
> 
> The linker fails on ESP32 (not S3).

This fails because the ESP32 platform has very limited IRAM available, and enabling PM capabilities pushes it beyond the limit. I’ll explore making the ESP-IDF components lighter to reduce memory usage without affecting Meshtastic functionality, but that may not be feasible. In the end, we may have to accept that dynamic light-sleep isn’t compatible with the ESP32.

In newer ESP-IDF versions (5.x compatible with Arduino 3.2.x), there’s a new sdkconfig option CONFIG_ESP_SYSTEM_ESP32_SRAM1_REGION_AS_IRAM, which makes more RAM available as IRAM, allowing the build to succeed. However, adopting this would be a breaking change for the project, since it requires moving from Arduino 2.x to 3.x. I’ve already created and tested such a build with the current Meshtastic version, but this shift is too significant to include in this PR.

m1nl commented:
> ```
> /work/.platformio/packages/toolchain-xtensa-esp32/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pio/build/heltec-v2_1/firmware.elf section `.iram0.text' will not fit in region `iram0_0_seg'
> /work/.platformio/packages/toolchain-xtensa-esp32/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: IRAM0 segment data does not fit.
> /work/.platformio/packages/toolchain-xtensa-esp32/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: region `iram0_0_seg' overflowed by 776 bytes
> collect2: error: ld returned 1 exit status
> ```
> 
> The linker fails on ESP32 (not S3).

This fails because the ESP32 platform has very limited IRAM available, and enabling PM capabilities pushes it beyond the limit. I’ll explore making the ESP-IDF components lighter to reduce memory usage without affecting Meshtastic functionality, but that may not be feasible. In the end, we may have to accept that dynamic light-sleep isn’t compatible with the ESP32.

In newer ESP-IDF versions (5.x compatible with Arduino 3.2.x), there’s a new sdkconfig option CONFIG_ESP_SYSTEM_ESP32_SRAM1_REGION_AS_IRAM, which makes more RAM available as IRAM, allowing the build to succeed. However, adopting this would be a breaking change for the project, since it requires moving from Arduino 2.x to 3.x. I’ve already created and tested such a build with the current Meshtastic version, but this shift is too significant to include in this PR.

m1nl commented on src/sleep.cpp:
```diff
@@ -68,30 +83,21 @@ RTC_DATA_ATTR int bootCount = 0;
-#if defined(ARCH_ESP32) && HAS_WIFI && !HAS_TFT
-
+#if defined(ARCH_ESP32) && !HAS_TFT
+#ifdef HAS_WIFI
```
sure, I'll check these definitions and I can update them to `bool`. I see there is inconsistency about how we use `HAS_XYZ` flags in the project. I will create a separate PR for this.

m1nl committed:
set serial clock source to ref_tick for esp32 platform

m1nl committed:
do not use ifdef for HAS_XYZ capability flags

m1nl committed:
switch heltec_v2.1 variant to esp-idf sdk with power management enabled

akohlsmith commented:
I'm testing this on ESP32C3 and so far the testing is looking very good. 

My only nitpick is that I think that there should be a comment in `variants/esp32c3/heltec_esp32c3/variant.h` (and others) which mentions the `HAS_32768HZ` macro and that it should be uncommented if your board has a 32kHz crystal. One of the big issues with supporting so many platforms is that these definitions tend to get buried in code.

e.g. 

```
// uncomment this is your board has a 32kHz crystal
//#define HAS_32768HZ 1
```

m1nl commented:
thx, you guys motivated me to rebase and update this PR this week; for now you can try to use https://github.com/m1nl/meshtastic-firmware/tree/esp_light_sleep_release_stable - it should **support** V4; I'll check and let you know

gxcreator commented:
Tried with power save mode, it was 10h+ instead of  ~5h on original firmware + 40MHz !

m1nl commented:
You can send me an email (check my GitHub profile). 

If you didn't try it before, please disable Bluetooth completely (in node settings) and check the current draw; I suspect it may be BT modem looking for your phone causing more frequent wake-ups.

iuvi7 commented:
<img width="181" height="168" alt="изображение" src="https://github.com/user-attachments/assets/3f4809a1-780e-4288-856f-af67ce5b4078" />

This one from Xiao Seed Studio quite popular as well, have few at shelf, will test it later by possibility from one 18650 :)

m1nl commented:
> > You can send me an email (check my GitHub profile).
> 
> Done. Main difference in logs with and without power safe enabling are too much `[GPS]` records on power safe mode. `[GPS]` records go about minute and this minute device consumes 110 mA. Then consumption decreases to 40 mA
> 
> PS: GPS is disabled in settings

From the logs, I can see that the system is actively searching for the GPS module but eventually stops because no GPS device is found. ("Disabled" in this context means that the system attempts to discover a GPS module but does not actually use it.)

Sleep functionality is quite limited while the device is searching for GPS, which could explain the increased power consumption. Was the GPS receiver disconnected from the board when power saving was enabled? If not, I’ll need to check whether the GPS module is being detected correctly when power saving mode is active.

phaseloop commented:
I have few questions:

- have you explored dynamic light sleep where entering sleep is managed by FreeRTOS scheduler based on scheduler queue and running idle task?

- does DFS in its current state interferes with regular full clock speed operations as it seems there is no code synchronization to indicate something needs full speed lock

- have you explored ModemSleep mode? According to Espressif docs, even without light sleep it's possible to go down to 20 mA power draw using combination of modem sleep settings and DFS without even officially entering sleep mode.

I suspect this whole power save architecture can be greatly simplified by not having separate sleep state and but rather sleeping in short intervals (like 100 ms) when rtos scheduler is idle and device have nothing better to do.

This is why NRF52 does not have officially a light sleep mode but works live in variable latency mode - doing internally something similar to what I described.

phaseloop commented:
Also - in your fork you say that this commit enables power management but it seems it added support for DSP (Digital Signal Processing):

https://github.com/espressif/arduino-esp32/commit/1685ef30f182428e0e76214083d108d123ee43b7

art-den commented:
> power difference between BT connected/disconnected state is expected as there are different frame timings involved

@phaseloop thanks for information

m1nl commented:
> I have few questions:
> 
> * have you explored dynamic light sleep where entering sleep is managed by FreeRTOS scheduler based on scheduler queue and running idle task?
> * does DFS in its current state interferes with regular full clock speed operations as it seems there is no code synchronization to indicate something needs full speed lock
> * have you explored ModemSleep mode? According to Espressif docs, even without light sleep it's possible to go down to 20 mA power draw using combination of modem sleep settings and DFS without even officially entering sleep mode.
> 
> I suspect this whole power save architecture can be greatly simplified by not having separate sleep state and but rather sleeping in short intervals (like 100 ms) when rtos scheduler is idle and device have nothing better to do.
> 
> This is why NRF52 does not have officially a light sleep mode but works live in variable latency mode - doing internally something similar to what I described.

1. Yes, that's what I'm using here (if dynamic light sleep is supported by Arduino framework)
2. There is only PM lock which prevents light-sleep. There is no PM lock for DFS, but I didn't really find any scenario when it should be used.
3. AFAIK modem-sleep is WiFi-specific and is used whenever WIFI_PS_MAX_MODEM is set; my PR sets MIN / MAX_MODEM depending on power save setting.

My PR is supposed to be compatible with both PM-enabled (customized) and regular (non-PM enabled) Arduino framework, so removing sleep states from PowerFSM is not a valid option.

m1nl committed:
reference CI-built framework package; add support for t-deck

m1nl committed:
setup gpio hold for buttons and keyboard

m1nl committed:
fix support for GC1109 PA (heltec v4)

m1nl committed:
fix power led handling when sleep is enabled

m1nl committed:
add support for tlora-pager

m1nl committed:
provide better compatibility with devices using default framework

m1nl committed:
use safe value for heap_caps_malloc_extmem_enable

current value of 256 bytes is very dangerous and may negatively
impact the overall performance; we need to keep in mind that spiram
is very slow comparing to regular RAM and many things may just fail;
in my case, dynamic light-sleep made the device unoperational; spiram
can be used to offload some costly allocations but this is not the
right way to handle that

thebentern commented:
@m1nl would your recommendation be to switch to 2048 for now?

thebentern committed:
Merge branch 'develop' into esp_light_sleep_release

cpatulea commented on variants/esp32s3/heltec_v3/platformio.ini:
```diff
@@ -1,10 +1,12 @@
+[env:heltec-v3]
 extends = esp32s3_base
+platform_packages =
+    platformio/framework-arduinoespressif32 @ https://github.com/m1nl/arduino-esp32/archive/refs/tags/2.0.17+5ae9873e.tar.gz ; disable WiFi IRAM optimizations in ESP-IDF
```
Sorry for the drive-by comment, I have been interested in esp32 migration (https://github.com/meshtastic/firmware/pull/9122) to pioandroid so that I can set some sdkconfig flags to solve a different problem (https://github.com/meshtastic/firmware/issues/6960).

pioarduino has a feature called 'hybrid compile' which allows to set a subset of sdkconfig options: https://deepwiki.com/pioarduino/platform-espressif32/2.9-sdk-configuration-and-build-flags

This can avoid the need to maintain a esp32-arduino-lib-builder repo, it would build arduino libs on-demand.

(though I found it might be limited and only allow control over a subset of sdk config options: https://github.com/meshtastic/firmware/issues/6960#issuecomment-3808930337)

Do you have any idea if pioarduino hybrid compile can provide the control you need for this?

I think avoiding to maintain a esp32-arduino-lib-builder could be a major benefit. I was planning on helping to port esp32 variants to pioarduino platform.

m1nl commented:
> > @thebentern I spent several days debugging why this PR made SPIRAM-capable devices extremely unstable when using latest develop branch. I was able to narrow the issue down to commit [5910cc2](https://github.com/meshtastic/firmware/commit/5910cc2e269cac13275d198434d34be210eb3d9d), which changes the SPIRAM allocation threshold to a very aggressive value of 256 bytes.
> > We need to keep in mind that SPIRAM is significantly slower than internal RAM, and moving small allocations there can easily cause subtle and hard-to-debug failures. In my case, enabling dynamic light sleep made the device completely unresponsive. With a 256-byte threshold, even very small buffers are pushed to SPIRAM, and this was causing Bluetooth to hang after sleep.
> 
> Wow, this is amazing debugging. I have run into similar issues on other projects but wasn't paying close enough attention here.
> 
> Does this PR allow I2C to come up early enough to enable the use of INAxxx devices to detect battery voltage for deep sleep? This was the biggest source of frustration in my earlier attempts at enabling light sleep for ESP32 and why my PRs were so difficult to manage. The devices I design do not use an ADC channel for battery voltage at all, and I had significant hacks to do be able to read the current/voltage sensor early enough.

Hey :) I think it does - when dynamic light sleep is enabled, all scheduled tasks will take place as requested. ESP-IDF will wake-up the device when next OSThread is to be run, so routines in Power.c should run as expected.

m1nl commented:
> One other question -- my own version of the espressif-arduino enabled PM for ESP32 (ESP32C3 only) also enables DFS and enables the use of an optional external 32.768kHz crystal. I took a brief look at your branch and didn't see either of these enabled. Is there any reason not to?

1. I think my code does enable DFS - take a look here https://github.com/m1nl/meshtastic-firmware/blob/1df8c24c02f1f13b7a0554d02461d653b09591d1/src/sleep.cpp#L502 .
2. 32.768kHz crystal is enabled for supported devices; there is a flag in `variant.h`, which enables that feature:
```
...
#define HAS_32768HZ 1
```

I believe it may be missing from several variants as defined in repo (i.e. T-Lora Pager, no schematic is available, but it can be enabled successfully).

m1nl commented:
@thebentern is there a chance you could help with the item below?

```
- my changes require ESP-IDF / Arduino with custom sdkconfig (https://github.com/m1nl/esp32-arduino-lib-builder) - would it be possible to move my repo under the meshtastic organization? I’ve added automated CI builds using GitHub Actions - everything is transparent now and no binary blobs are hosted in the repository.
```

I can set up the CI pipeline and migrate the repository contents. After that, I’ll update the PR so all links to the custom Arduino framework point to releases under the Meshtastic GitHub organization.

akohlsmith commented:
I do not see light sleep getting enabled, only periodic AGC reset attempts and messages about gpio_hold_en being attempted on output-capable IOs:

```
[2026-02-25 16:00:41] DEBUG | ??:??:?? 104 [PowerFSM] Setup radio interrupt (GPIO03) with wakeup by GPIO interrupt
[2026-02-25 16:00:42] E (106235) gpio: gpio_hold_en(641): Only output-capable GPIO support this function
[2026-02-25 16:00:42] E (106238) gpio: gpio_hold_en(641): Only output-capable GPIO support this function
[2026-02-25 16:00:42] E (106245) gpio: gpio_hold_en(641): Only output-capable GPIO support this function
[2026-02-25 16:00:42] E (106252) gpio: gpio_hold_en(641): Only output-capable GPIO support this function
[2026-02-25 16:00:42] E (106260) gpio: gpio_hold_en(641): Only output-capable GPIO support this function
[2026-02-25 16:01:41] DEBUG | ??:??:?? 164 attempting AGC reset
[2026-02-25 16:01:41] DEBUG | ??:??:?? 224 attempting AGC reset
[2026-02-25 16:02:41] DEBUG | ??:??:?? 284 attempting AGC reset
[2026-02-25 16:03:41] DEBUG | ??:??:?? 344 attempting AGC reset
```

I do have light sleep enabled in the configuration:

```
$ meshtastic --port /dev/cu.usbserial-14101 --get power
Connected to radio
power.is_power_saving: True
power.wait_bluetooth_secs: 60
power.sds_secs: 3600
power.ls_secs: 86400
power.min_wake_secs: 60
power.device_battery_ina_address: 64
Completed getting preferences
```

I also do not see the "Using INA on I2C addr 0x%x for charging detection" message on bootup, although I don't see why I shouldn't: I have an INA219, I have telemetry enabled (I see those messages in the log and being transmitted over LoRa), I'm not excluding environmental sensing and `DISABLE_INA_CHARGING_DETECTION` isn't defined anywhere.

akohlsmith commented:
it looks like the AGC reset seems to never re-enter light sleep -- every minute it attempts an AGC reset, but I never see `doLightSleep()` get called until after a packet is received:

```
[2026-02-25 16:27:00] INFO  | ??:??:?? 104 [PowerFSM] ABK doLightSleep(39146 ms)
[2026-02-25 16:27:00] DEBUG | ??:??:?? 104 [PowerFSM] Setup radio interrupt (GPIO03) with wakeup by GPIO interrupt
[2026-02-25 16:27:00] DEBUG | ??:??:?? 164 attempting AGC reset
[2026-02-25 16:28:00] DEBUG | ??:??:?? 224 attempting AGC reset
[2026-02-25 16:29:00] DEBUG | ??:??:?? 284 attempting AGC reset
[2026-02-25 16:30:00] DEBUG | ??:??:?? 344 attempting AGC reset
[2026-02-25 16:31:00] DEBUG | ??:??:?? 404 attempting AGC reset
```

I'm digging into this a little more but wanted to mention it in case you might already know what's going on

cpatulea commented:
Discord thread: https://discord.com/channels/867578229534359593/1476361026801762508

gxcreator commented:
I've ported this to current develop branch and with few tweaks(fixed Wi-Fi PS which is major gain, saved ~80 mA and few light-sleep tweaks) I could reach **~20-30mA** while Wi-Fi connected on Heltec Wireless Paper - while original Meshtastic draws 160 mA! 

@m1nl  patch lowered 160 -> ~120
<img width="336" height="230" alt="Screenshot_20260830_193134" src="https://github.com/user-attachments/assets/f1be5f72-44c5-4827-9191-fcd31ceb3f47" />

Properly enabling auto light-sleep and DFS:
<img width="336" height="230" alt="Screenshot_20260830_212109" src="https://github.com/user-attachments/assets/507cd820-a343-46cd-a425-ae8589abc24a" />

Enabling Wi-fi PS:
<img width="340" height="234" alt="Screenshot_20260830_222241" src="https://github.com/user-attachments/assets/9b06da1b-87b9-4296-81af-687caa62b912" />

Some light-sleep tweaks:

<img width="345" height="238" alt="Screenshot_20260831_001303" src="https://github.com/user-attachments/assets/25f9944b-9a5b-4e42-9cf2-f80495e90924" />


Looks very promising, almost 10x battery life! I still need to sort out light sleep control since meshtastic impl is a bit shaky and then push to another PR probably

Thanks @m1nl for initial work and inspiration!

m1nl commented:
@gxcreator great; I didn't have time to move it forward (after a few months of rebasing this PR); did you make use of sdkconfig via PlatformIO ? as you can see I was using a custom ESP-IDF + Arduino build and this was a blocker at the time my changes were discussed; I'd be really happy to see your code / PR and can help fixing stuff if needed

gxcreator commented:
> @gxcreator great; I didn't have time to move it forward (after a few months of rebasing this PR); did you make use of sdkconfig via PlatformIO ? as you can see I was using a custom ESP-IDF + Arduino build and this was a blocker at the time my changes were discussed; I'd be really happy to see your code / PR and can help fixing stuff if needed

I use just
```
custom_sdkconfig =
  ${esp32s3_base.custom_sdkconfig}
  CONFIG_PM_ENABLE=y
  CONFIG_FREERTOS_USE_TICKLESS_IDLE=y

```
in `variants/esp32s3/heltec_wireless_paper/platformio.ini`

m1nl commented:
> > @gxcreator great; I didn't have time to move it forward (after a few months of rebasing this PR); did you make use of sdkconfig via PlatformIO ? as you can see I was using a custom ESP-IDF + Arduino build and this was a blocker at the time my changes were discussed; I'd be really happy to see your code / PR and can help fixing stuff if needed
> 
> I use just
> 
> ```
> custom_sdkconfig =
>   ${esp32s3_base.custom_sdkconfig}
>   CONFIG_PM_ENABLE=y
>   CONFIG_FREERTOS_USE_TICKLESS_IDLE=y
> ```
> 
> in `variants/esp32s3/heltec_wireless_paper/platformio.ini`

great - looking forward to see your code then
